topostats/tracing/disordered_tracing.py

```diff
@@ -0,0 +1,708 @@
+) -> tuple[dict[int, npt.NDArray], dict[int, npt.NDArray]]:
+    """
+    Take an image and labelled mask and crops individual grains and original heights to a list.
+
+    A second padding is made after cropping to ensure for "edge cases" where grains are close to bounding box edges that
+    they are traced correctly. This is accounted for when aligning traces to the whole image mask.
+
+    Parameters
+    ----------
+    image : npt.NDArray
+        Gaussian filtered image. Typically filtered_image.images["gaussian_filtered"].
+    labelled_grains_mask : npt.NDArray
+        2D Numpy array of labelled grain masks, with each mask being comprised solely of unique integer (not
+        zero). Typically this will be output from 'grains.directions[<direction>["labelled_region_02]'.
+    pad_width : int
+        Cells by which to pad cropped regions by.
+
+    Returns
+    -------
+    Tuple
+        Returns a tuple of two dictionaries, each consisting of cropped arrays.
+    """
+    # Get bounding boxes for each grain
+    region_properties = skimage_measure.regionprops(labelled_grains_mask)
+    # Subset image and grains then zip them up
+    cropped_images = {}
+    cropped_masks = {}
+
+    # for index, grain in enumerate(region_properties):
+    #    cropped_image, cropped_bbox = crop_array(image, grain.bbox, pad_width)
+
+    cropped_images = {index: crop_array(image, grain.bbox, pad_width) for index, grain in enumerate(region_properties)}
+    cropped_images = {index: np.pad(grain, pad_width=pad_width) for index, grain in cropped_images.items()}
+    cropped_masks = {
+        index: crop_array(labelled_grains_mask, grain.bbox, pad_width) for index, grain in enumerate(region_properties)
+    }
+    cropped_masks = {index: np.pad(grain, pad_width=pad_width) for index, grain in cropped_masks.items()}
+    # Flip every labelled region to be 1 instead of its label
+    cropped_masks = {index: np.where(grain == 0, 0, 1) for index, grain in cropped_masks.items()}
+    # Get BBOX coords to remap crops to images
+    bboxs = [pad_bounding_box(image.shape, list(grain.bbox), pad_width=pad_width) for grain in region_properties]
+
+    return (cropped_images, cropped_masks, bboxs)
+
+
+def grain_anchor(array_shape: tuple, bounding_box: list, pad_width: int) -> list:
+    """
+    Extract anchor (min_row, min_col) from labelled regions and align individual traces over the original image.
+
+    Parameters
+    ----------
+    array_shape : tuple
+        Shape of original array.
+    bounding_box : list
+        A list of region properties returned by 'skimage.measure.regionprops()'.
+    pad_width : int
+        Padding for image.
+
+    Returns
+    -------
+    list(Tuple)
+        A list of tuples of the min_row, min_col of each bounding box.
+    """
+    bounding_coordinates = pad_bounding_box(array_shape, bounding_box, pad_width)
+    return (bounding_coordinates[0], bounding_coordinates[1])
+
+
+def trace_grain(
```

Comment:
Was moving this intentional?

`trace_grain()` and related functions are something that I thought was to be done _after_ getting a disordered trace (by either of the two methods). I would have thought this should have remained in `dnatracing.py`. :thinking:

Reply:
Yes as it's similar to how this was initially done in `dnatracing.py`, i.e. each grain is individually processed to get the disordered traces. Then the whole image is built up from the individual crops and this is the result. Then the next step will take the full pruned image and individually prcess each disordered trace to produce nodestats/ordered traces with it.

There was an initial thought to do this inside the loop of a single grain but we decided against this when we spoke in your office.

Reply:
This has been renamed to "disordered_trace_grain()"